### DISCUSSION TITLE
freeze when returning a component from a function

### DISCUSSION THREAD

**Discussion by uyha-kwz** [ArthurSonzogni/FTXUI#664] [Q&A] [answered]
I want to have a function that returns a component, I have the function defined as follows:
```cpp
template <typename T>
auto number(T &source) -> ftxui::Component {
  using namespace ftxui;

  auto decrease  = Button("-", [&] { --source; });
  auto increase  = Button("+", [&] { ++source; });
  auto container = Container::Horizontal({decrease, increase});
  return Renderer(container, [&] {
    return hbox({
        decrease->Render(),
        text(fmt::format("{}", source)) | center,
        increase->Render(),
    });
  });
}
```

then I have the following snippet in main
```cpp
int main() {
  using namespace ftxui;

  auto level    = std::uint8_t{0};
  auto renderer = number(level);
  auto screen   = ScreenInteractive::Fullscreen();

  screen.Loop(renderer);
}
```
but then the application just freezes when I run it, what am I doing wrong here?

**Comment by ArthurSonzogni** [comment, marked as answer]
Hello @uyha-kwz

Your lambda captures `decrease` and `increase` declared from the stack by reference. It is not valid using them after the end of the function.

Your binary suffers from undefined behavior / memory safety issues.

### Capture the Component (`std::shared_ptr`) by value in the lambda. (Not recommanded)

```cpp
template <typename T>
auto number(T &source) -> ftxui::Component {
  using namespace ftxui;

  auto decrease  = Button("-", [&] { --source; });
  auto increase  = Button("+", [&] { ++source; });
  auto container = Container::Horizontal({decrease, increase});
  return Renderer(container, [decrease, increase, &source] {
    return hbox({
        decrease->Render(),
        text(std::to_string(source)) | center,
        increase->Render(),
    });
  });
}

```

### Implement your own `ftxui::ComponentBase` and store them in.

```cpp
template <typename T>
auto number(T& source) -> ftxui::Component {
  using namespace ftxui;

  class Number : public ComponentBase {
   private:
    T& source_;
    Component decrease_ = Button("-", [&] { --source_; });
    Component increase_ = Button("+", [&] { ++source_; });
    Component container_ = Container::Horizontal({
        decrease_,
        increase_,
    });

   public:
    Number(T& source) : source_(source) { Add(container_); }

    Element Render() override {
      return hbox({
          decrease_->Render(),
          text(std::to_string(source_)) | center,
          increase_->Render(),
      });
    }
  };

  return Make<Number>(source);
}

```

**Reply by uyha-kwz** [reply]
oh, I was assuming that since `Button` is a `Component` which is a `std::shared_ptr<ComponentBase>`. Hence I assumed that the `Renderer` will hold a reference to it which prevent the underlying object being deallocated, but apparently, that's not the case.

**Reply by ArthurSonzogni** [reply]
Yes, that's tricky.

The `shared_ptr<ComponentBase>` will be copied around in the tree, so the `ComponentBase` will be kept alive.

However we are tacking a reference to `decrease` and `increase`. This translates to `shared_ptr<T>&`. The shared_ptr is "dead" after the end of the function.

It is a case where you have `T`, but you don't have `shared_ptr<T>` anymore.

**Reply by uyha-kwz** [reply]
oh, they are being passed as references, I thought they were being copied. Thanks a lot for the explanation. This is when I wish C++ has a more explicit syntax for passing reference like Rust.

**Reply by ArthurSonzogni** [reply]
You can pass:
- by **copy** using `[argument]`
- by **reference** using `[&argument]`
- by **move** using `[argument = std::move(x)]`

Then the `[&]` and `[=]` shorthands means capturing everything else by **reference**, and by **copy** respectively.

**Reply by uyha-kwz** [reply]
ok, that makes sense. Can you elaborate why capturing by value is not recommended? As far as I can tell, it's similar to the implementation by hand.

**Reply by ArthurSonzogni** [reply]
Capturing by value is great in general.

Here, I am not recommanding it, because when the object to copy is a `std::shared_ptr`, it might be tricky to understand we copy the pointer and not object.

You can still use it if you want and understand what happens.

**Reply by uyha-kwz** [reply]
thank you very much for the answers.
